ports/nrf/common-hal/busio/UART.c

```diff
@@ -35,58 +35,329 @@
+    _VERIFY_ERR(*errcode);
+    (*errcode) = 0;
+
+    while ( nrfx_uarte_tx_in_progress(&self->uarte) && (ticks_ms - start_ticks < self->timeout_ms) ) {
+#ifdef MICROPY_VM_HOOK_LOOP
+        MICROPY_VM_HOOK_LOOP
+#endif
+    }
+
+    if ( !nrfx_is_in_ram(data) ) {
+        gc_free(tx_buf);
+    }
+
+    return len;
 }
 
 uint32_t common_hal_busio_uart_get_baudrate(busio_uart_obj_t *self) {
-    mp_raise_NotImplementedError(translate("busio.UART not yet implemented"));
     return self->baudrate;
 }
 
 void common_hal_busio_uart_set_baudrate(busio_uart_obj_t *self, uint32_t baudrate) {
-    mp_raise_NotImplementedError(translate("busio.UART not yet implemented"));
     self->baudrate = baudrate;
+    nrf_uarte_baudrate_set(self->uarte.p_reg, get_nrf_baud(baudrate));
 }
 
 uint32_t common_hal_busio_uart_rx_characters_available(busio_uart_obj_t *self) {
+    return (self->rx_count > 0) ? self->rx_count : 0;
+}
+
+void common_hal_busio_uart_clear_rx_buffer(busio_uart_obj_t *self) {
+
+}
+
+bool common_hal_busio_uart_ready_to_tx(busio_uart_obj_t *self) {
+    return !nrfx_uarte_tx_in_progress(&self->uarte);
+}
+
+static uint32_t get_nrf_baud (uint32_t baudrate)
+{
+    if ( baudrate <= 1200 ) {
+        return NRF_UARTE_BAUDRATE_1200;
+    }
+    else if ( baudrate <= 2400 ) {
+        return NRF_UARTE_BAUDRATE_2400;
+    }
+    else if ( baudrate <= 4800 ) {
+        return NRF_UARTE_BAUDRATE_4800;
+    }
+    else if ( baudrate <= 9600 ) {
+        return NRF_UARTE_BAUDRATE_9600;
+    }
+    else if ( baudrate <= 14400 ) {
+        return NRF_UARTE_BAUDRATE_14400;
+    }
+    else if ( baudrate <= 19200 ) {
+        return NRF_UARTE_BAUDRATE_19200;
+    }
+    else if ( baudrate <= 28800 ) {
+        return NRF_UARTE_BAUDRATE_28800;
+    }
+    else if ( baudrate <= 38400 ) {
+        return NRF_UARTE_BAUDRATE_38400;
+    }
+    else if ( baudrate <= 57600 ) {
+        return NRF_UARTE_BAUDRATE_57600;
+    }
+    else if ( baudrate <= 76800 ) {
+        return NRF_UARTE_BAUDRATE_76800;
+    }
+    else if ( baudrate <= 115200 ) {
+        return NRF_UARTE_BAUDRATE_115200;
+    }
+    else if ( baudrate <= 230400 ) {
+        return NRF_UARTE_BAUDRATE_230400;
+    }
+    else if ( baudrate <= 250000 ) {
+        return NRF_UARTE_BAUDRATE_250000;
+    }
+    else if ( baudrate <= 460800 ) {
+        return NRF_UARTE_BAUDRATE_460800;
+    }
+    else if ( baudrate <= 921600 ) {
+        return NRF_UARTE_BAUDRATE_921600;
+    }
+    else {
+        return NRF_UARTE_BAUDRATE_1000000;
+    }
+}
+
+#else
+
+void common_hal_busio_uart_construct (busio_uart_obj_t *self,
+                                      const mcu_pin_obj_t * tx, const mcu_pin_obj_t * rx, uint32_t baudrate,
+                                      uint8_t bits, uart_parity_t parity, uint8_t stop, uint32_t timeout,
+                                      uint8_t receiver_buffer_size) {
+    mp_raise_NotImplementedError(translate("busio.UART not yet implemented"));
```

Comment:
I think these should all be "not available" instead of "not yet implemented", because on the nRF52832 there will never be another UART instance, right?

Reply:
ah right, I will change it and update the new string